src/xb-silo.c

```diff
@@ -936,21 +928,7 @@ xb_silo_new_from_xml (const gchar *xml, GError **error)
 XbNode *
 xb_silo_node_create (XbSilo *self, XbSiloNode *sn)
 {
-	XbNode *n;
-	XbSiloPrivate *priv = GET_PRIVATE (self);
-	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->nodes_mutex);
-
-	g_return_val_if_fail (locker != NULL, NULL);
-
-	/* does already exist */
-	n = g_hash_table_lookup (priv->nodes, sn);
-	if (n != NULL)
-		return g_object_ref (n);
```

Comment:
hmm, I added this code for a reason, perhaps for fwupd. Could we conditionalize it with a flag perhaps?

Reply:
I think I'd prefer a flag, default caching. GNOME Software can set _NO_NODE_CACHE or something.

Reply:
I’ve read through the code in fwupd in more detail, and I don’t think there’s a neat way to avoid this caching without massively breaking fwupd’s API and reorganising how it passes releases around internally. Basically, the CAB parser is sideloading three bits of binary data onto the `XbNode` for each `release` in its appstream, that’s getting passed around through some other code, and then the engine is reading that sideloaded data and storing it in a `FuRelease` object. If the CAB loader created the `FuRelease` object in the first place, the sideloading wouldn’t be necessary. Essentially the issue is that `XbNode` is being used to expose the XML structure, but also to represent the higher level object (a release) that’s described by the XML. I don’t think it’s possible to be efficient easily with that wide a range of possible use cases. In this case, libxmlb is not “doing one thing and doing it well”. :(

I’ll look into adding a flag. I think it’s a hack, but it seems unavoidable if I’m understanding this fwupd code correctly.

Reply:
@hughsie, can you explain your reasoning for preferring a flag please? I’d like to make sure I understand the use cases and considerations at work here, in case they affect future optimisation work I might do. :)

Reply:
> can you explain your reasoning for preferring a flag please

That while I think doing things that require lockstep changes to gnome-software and libxmlb is probably an acceptable compromise (e.g. the API break for speed/memory) I think breaking 2/2 consumers is probably going too far.

Reply:
> breaking 2/2 consumers is probably going too far.

(The original version of this only broke fwupd, not gnome-software.)

I’ve pushed a new version which adds some new flags and properties to the API to make the cache conditional. It’s a bit fiddly how the thread safety interacts with whether the cache is enabled/disabled and I’m not entirely happy with it. Open to suggestions for improving it.

Reply:
> It’s a bit fiddly how the thread safety interacts with whether the cache is enabled/disabled and I’m not entirely happy with it.

One option for improving that would be to add a construct-only `enable-node-cache` property to `XbBuilder`, which can be used to ensure that the `XbSilo` is always *constructed* with the correct `enable-node-cache` value (it currently can’t be, since it is constructed in `xb_builder_init()`, which is not where the `NO_NODE_CACHE` flag is available). Then `XbSilo:enable-node-cache` can be made construct-only, and the thread safety concerns go away.